javalib/src/main/scala/java/lang/String.scala

```diff
@@ -1,42 +1,890 @@
-  def indexOf(ch: scala.Int, fromIndex: scala.Int): scala.Int = ???
-  def lastIndexOf(ch: scala.Int): scala.Int = ???
-  def lastIndexOf(ch: scala.Int, fromIndex: scala.Int): scala.Int = ???
-  def substring(beginIndex: scala.Int): _String = ???
-  def substring(beginIndex: scala.Int, endIndex: scala.Int): _String = ???
-  def startsWith(prefix: _String): scala.Boolean = ???
-  def toLowerCase(): _String = ???
-  def toCharArray(): Array[Char] = ???
-  def equalsIgnoreCase(str: _String): scala.Boolean = ???
-  def getChars(srcBegin: scala.Int, srcEnd: scala.Int, dst: Array[scala.Char],
-      dstBegin: scala.Int): Unit = ???
-  override def hashCode(): scala.Int = ???
-
-  // TODO: rest of the api
+import java.io.{Serializable, UnsupportedEncodingException}
+import java.util._
+import java.util.regex._
+import java.nio._
+import java.nio.charset._
+
+final class _String private ()
+    extends Serializable with Comparable[_String] with CharSequence {
+  private var value         = new Array[Char](0)
+  private var offset        = 0
+  private var count         = 0
+  private var cashedHashCode: Int = _
+
+  def this(data: Array[scala.Byte], _high: Int, start: Int, length: Int) = {
+    this()
+    if (length <= data.length - start && start >= 0 && 0 <= length) {
+      offset = 0
+      value = new Array[Char](length)
+      var high = _high
+      high <<= 8
+
+      var i = 0
+      while (i < length) {
+        value(i) = (high + data(start + i) & 0xff).toChar
+        i += 1
+      }
+    } else {
+      throw new StringIndexOutOfBoundsException()
+    }
+  }
+
+  def this(
+      data: Array[scala.Byte], start: Int, length: Int, encoding: Charset) = {
+    this()
+    if (encoding == null) {
+      throw new NullPointerException()
+    } else {
+      offset = 0
+      _String.lastCharset = encoding
+
+      val charBuffer = encoding.decode(ByteBuffer.wrap(data, start, length))
+      value = charBuffer.array
+      count = charBuffer.length
+    }
+  }
+
+  def this(
+      data: Array[scala.Byte], start: Int, length: Int, encoding: _String) =
+    this(data, start, length, Charset.forName(encoding))
+
+  def this(data: Array[scala.Byte], start: Int, length: Int) =
+    this(data, start, length, Charset.defaultCharset())
+
+  def this(data: Array[scala.Byte], high: Int) =
+    this(data, high, 0, data.length)
+
+  def this(data: Array[scala.Byte], encoding: _String) =
+    this(data, 0, data.length, encoding)
+
+  def this(data: Array[scala.Byte], encoding: Charset) =
+    this(data, 0, data.length, encoding)
+
+  def this(data: Array[scala.Byte]) =
+    this(data, 0, data.length)
+
+  def this(data: Array[Char], start: Int, length: Int) = {
+    this()
+    if (start >= 0 && 0 <= length && length <= data.length - start) {
+      offset = 0
+      value = new Array[Char](length)
+      count = length
+      System.arraycopy(data, start, value, 0, count)
+    } else {
+      throw new StringIndexOutOfBoundsException()
```

Comment:
This should be an `IndexOutOfBoundsException`. See http://docs.oracle.com/javase/8/docs/api/java/lang/String.html#String-char:A-int-int-